### PR TITLE
Blockchain database validators

### PR DESCRIPTION
**Pull request by CjS77**
We provde a set of `Validators` to `BlockchainDatabase` so that
it can test whether new blocks are valid.

We use a dedicated Validator trait rather than embedding functions
into the DB because it makes testing easier; this PR being a case
in point. Each time we update the validation rules, all the unit
tests for other aspects of the blockchain that make use of a
test DB would need to be updated. For this reason, `MockValidator`
is used everywhere to allow the current test suite to pass.

The actual Base Node executable itself uses a correct set of
Validators.

The validators are still missing some steps, which have been
marked in the code, with correspondinng issues assigned to
the TODOs.

### CODE REVIEW

File: base_layer/core/src/validation/block_validators.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+impl StatelessValidator {
+    pub fn new(rules: Arc<ConsensusRules>, factories: Arc<CryptoFactories>) -> Self {
+        Self { rules, factories }
+    }
+}
+
+impl<B: BlockchainBackend> Validation<Block, B> for StatelessValidator {
+    /// The consensus checks that are done (in order of cheapest to verify to most expensive):
+    /// 1. Is there precisely one Coinbase output and is it correctly defined?
+    /// 1. Is the accounting correct?
```
Comment (neonknight64) [inline base_layer/core/src/validation/block_validators.rs]:
```suggestion
    /// 2. Is the accounting correct?
```

File: base_layer/core/src/validation/block_validators.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    /// 1. Are all inputs allowed to be spend (Are the feature flags satisfied)
```
Comment (neonknight64) [inline base_layer/core/src/validation/block_validators.rs]:
```suggestion
    /// 3. Are all inputs allowed to be spend (Are the feature flags satisfied)
```

File: base_layer/core/src/validation/block_validators.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+}
+
+impl<B: BlockchainBackend> Validation<Block, B> for FullConsensusValidator<B> {
+    /// The consensus checks that are done (in order of cheapest to verify to most expensive):
+    /// 1. Does the block satisfy the stateless checks?
+    /// 1. Are all inputs currently in the UTXO set?
+    /// 1. Is the block header timestamp within range?
+    /// 1. Is the Proof of Work valid?
+    /// 1. Is the achieved difficulty of this block >= the target difficulty for this block?
```
Comment (neonknight64) [inline base_layer/core/src/validation/block_validators.rs]: The numbering seems to be incorrect.

File: base_layer/core/src/validation/block_validators.rs (hunk 1 above)
Comment (sdbondi) [reply base_layer/core/src/validation/block_validators.rs]: This is correct markdown (see comment source)

1. "1."
1. means
1. a
1. numbered 
1. list 
1. in
1. markdown

Comment (neonknight64) [reply base_layer/core/src/validation/block_validators.rs]: I see, that is Cool.

**Pull request merged**